pennylane/beta/tapes/rev.py

```diff
@@ -0,0 +1,204 @@
+            vec1_indices=vec1_indices,
+            obs_indices=obs_indices,
+            vec2_indices=vec2_indices,
+        )
+
+        return device._einsum(einsum_str, device._conj(vec1), mat, vec2)
+
+    def jacobian(self, device, params=None, **options):
+        # The parameter_shift_var method needs to evaluate the circuit
+        # at the unshifted parameter values; the pre-rotated statevector is then stored
+        # self._state attribute. Here, we set the value of the attribute to None
+        # before each Jacobian call, so that the statevector is calculated only once.
+        self._state = None
+        return super().jacobian(device, params, **options)
+
+    def analytic_pd(self, idx, device, params, **options):
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        # The reversible tape only support differentiating
+        # expectation values of observables for now.
+        for m in self.measurements:
+            if (
+                m.return_type is qml.operation.Variance
+                or m.return_type is qml.operation.Probability
+            ):
+                raise ValueError(
+                    f"{m.return_type} is not supported with the reversible gradient method"
+                )
+
+        # The reversible tape only supports the RX, RY, RZ, and Rot operations for now:
+        #
+        # * CRX, CRY, CRZ ops have a non-unitary matrix as generator.
+        #
+        # * PauliRot, MultiRZ, U2, and U3 do not have generators specified.
+        #
+        # TODO: the controlled rotations can be supported by multiplying ``state``
+        # directly by these generators within this function
+        # (or by allowing non-unitary matrix multiplies in the simulator backends)
+
+        if op.name not in ["RX", "RY", "RZ", "Rot"]:
+            raise ValueError(
+                "The {} gate is not currently supported with the "
+                "reversible gradient method.".format(op.name)
+            )
+
+        if self._state is None:
+            self.execute_device(params, device)
+            self._state = device._pre_rotated_state
+
+        self.set_parameters(params)
+
+        # create a new circuit which rewinds the pre-measurement state to just after `op`,
+        # applies the generator of `op`, and then plays forward back to
+        # pre-measurement step
+        wires = op.wires
+        op_idx = self.operations.index(op)
+
+        # TODO: likely better to use circuitgraph to determine minimally necessary ops
+        between_ops = self.operations[op_idx + 1 :]
+
+        if op.name == "Rot":
+            decomp = op.decomposition(*op.parameters, wires=wires)
+            generator, multiplier = decomp[p_idx].generator
+            between_ops = decomp[p_idx + 1 :] + between_ops
+        else:
+            generator, multiplier = op.generator

```

Comment:
this could likely be modified to take advantage of the new tape structure, but it's not entirely clear to me how

Reply:
Could use `expand` instead of `decomposition`

Reply:
I almost did! My fear is that `expand()` has more overhead